### PR TITLE
Forwarding random port with upnpc #700

### PR DESCRIPTION
**Pull request by apoorv-2204**
# Description

We chose a random port , from a list of unassigned  ports as per IANA.

Fixes 

- [ ] #700 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Manual
image

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/networking/port_forwarding.ex [hunk 1]
```diff
@@ -70,4 +75,27 @@ defmodule Archethic.Networking.PortForwarding do
   defp ip_lookup_provider do
     Application.get_env(:archethic, IPLookup)
   end
+
+  @iana_registry_path Path.join([
```
Comment (samuelmanzanera) [inline lib/archethic/networking/port_forwarding.ex:79]: You should put non mutable (by code) resources in the priv folder .

File: lib/archethic/networking/port_forwarding.ex [hunk 2]
```diff
@@ -70,4 +75,27 @@ defmodule Archethic.Networking.PortForwarding do
...
+                        File.cwd() |> elem(1),
+                        "lib",
+                        "archethic",
+                        "networking",
+                        "registry",
+                        "iana_unassigned_ports"
+                      ])
+
+  # see https://github.com/apoorv-2204/iana_port_registry
+  defp list_iana_unassigned_ports() do
```
Comment (samuelmanzanera) [inline lib/archethic/networking/port_forwarding.ex:89]: I think you could simplify without the need to list the port from IANA. Between 49152 through 65535 should make it. 
The randomness could be recursive until one port is free.

**Review by samuelmanzanera — CHANGES_REQUESTED** [review]

**Review by samuelmanzanera — COMMENTED** [review]
Code seems ok for me

**Review by samuelmanzanera — APPROVED** [review]

**Comment by bchamagne** [comment]
I see some functions that use this code are completely ignoring the random port:

ex:  Archethic.Networking.Scheduler

```
  defp open_ports do
    p2p_port = Application.get_env(:archethic, P2PListener) |> Keyword.fetch!(:port)
    web_port = Application.get_env(:archethic, WebEndpoint) |> get_in([:http, :port])

    with {:ok, _} <- PortForwarding.try_open_port(p2p_port, false),
         {:ok, _} <- PortForwarding.try_open_port(web_port, false) do
      {:ok, p2p_port, web_port}
    end
  end
```

**Comment by bchamagne** [comment]
what about HTTP(S) port, you don't want to open random port for them (ex: ArchethicWeb.Supervisor.init)

**Comment by bchamagne** [comment]
> what about HTTP(S) port, you don't want to open random port for them (ex: ArchethicWeb.Supervisor.init)

Not an issue, I didn't notice the `force = false` flag.

**Review by samuelmanzanera — APPROVED** [review]

**Comment by samuelmanzanera** [comment]
@apoorv-2204 Could you check why the CI is not passing ?

**Comment by apoorv-2204** [comment]
> @apoorv-2204 Could you check why the CI is not passing ?

Thanks for pointing it out

**Review by Neylix — COMMENTED** [review]
On line 21 of PortForwarding module, if the function `do_try_open_port` return an error, the fallback is never called so the random port is never published

**Comment by apoorv-2204** [comment]
> On line 21 of PortForwarding module, if the function `do_try_open_port` return an error, the fallback is never called so the random port is never published

```  {:ok, port} <- do_try_open_port(port) do
      {:ok, port}```

**Review by Neylix — APPROVED** [review]

**Pull request merged**